File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        i = _max = 0
        seen = set()
        for j in range(len(s)):
            c = s[j]
            while c in seen:
                seen.remove(s[i])
                i += 1
            seen.add(s[j])
            _max = max(_max, j-i+1)
        return _max
            
        

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> seen = new HashSet<Character>();
        int j = 0, res = 0;
        for(int i = 0; i < s.length(); i++){
            char x = s.charAt(i);
            while(seen.contains(x)){
                seen.remove(s.charAt(j));
                j += 1;
            }
            seen.add(s.charAt(i));
            res = Math.max(res, i-j+1);
        }
        return res;
    }
}